### DISCUSSION TITLE
Is there a recommended pattern to ensure finalizer functions are invoked when cleaning up the Runtime/Context?

### DISCUSSION THREAD

**Discussion by TooTallNate** [quickjs-ng/quickjs#174] [Q&A]
When using the "Class" API in QuickJS (`JS_NewClass()`), I'm seeing the finalizer function not be invoked when the program is shutting down, which leads to reported memory leaks. I would want these finalizers to be invoked before the `JS_FreeContext(ctx); JS_FreeRuntime(rt);` calls. Is this the normal behavior of QuickJS or am I doing something wrong?

**Comment by bnoordhuis** [comment]
JS_FreeRuntime() should run any finalizers whose objects have a reference count of 1. Try building quickjs with `-DDUMP_LEAKS=1`.

**Reply by TooTallNate** [reply]
Sure, so for example I get something like:

```
Object leaks:
       ADDRESS REFS SHRF          PROTO      CLASS PROPS
   0xa720d8f80    1   0*            0x0 TlsContext {  }
Memory leak: 1160 bytes lost in 4 blocks
```

So since REFS is 1, should that not qualify for finalization?

**Reply by saghul** [reply]
Is that held on another JS object but on the native side by any chance?

Then you'll need to mark it in the mark phase too.
